File: server_python/routes/scan.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from datetime import datetime
from .. import models, schemas, database

router = APIRouter()

@router.post("/")
@router.post("")
def scan_qr(request: schemas.ScanRequest, db: Session = Depends(database.get_db)):
    # 1. Input Validation
    if not request.roll_no or not request.event_id:
        raise HTTPException(status_code=400, detail="roll_no and event_id are required")
    
    # Lyra Mode: Strict String Ingestion (No transformations)
    clean_roll_no = request.roll_no.strip().upper() 
    
    # 2. Check Event
    event = db.query(models.Event).filter(models.Event.id == request.event_id).first()
    if not event:
        raise HTTPException(status_code=404, detail="Event not found")
    if not event.is_active:
        raise HTTPException(status_code=400, detail="Event is no longer active")

    # 3. Check Student
    from sqlalchemy import func
    student = db.query(models.Student).filter(func.lower(models.Student.roll_no) == clean_roll_no.lower()).first()
    if not student:
        print(f"❌ Scan attempt failed: Student {clean_roll_no} not found")
        raise HTTPException(status_code=404, detail=f"No student registered with roll number: {clean_roll_no}")

    # 4. Check Existing Log
    log = db.query(models.AttendanceLog).filter(
        models.AttendanceLog.roll_no == clean_roll_no,
        models.AttendanceLog.event_id == request.event_id
    ).first()

    now = datetime.now()
    now_iso = now.isoformat()

    if not log:
        # === FIRST SCAN -> CHECK-IN ===
        new_log = models.AttendanceLog(
            roll_no=clean_roll_no,
            event_id=request.event_id,
            check_in_time=now_iso,
            status="PENDING"
        )
        db.add(new_log)
        db.commit()
        db.refresh(new_log)
        
        return {
            "action": "CHECK_IN",
            "roll_no": clean_roll_no,
            "student_name": student.name,
            "check_in_time": now_iso,
            "message": f"✅ {student.name} checked in successfully"
        }
    
    else:
        # Check for rapid re-scan (Anti-Abuse)
        last_action_time = datetime.fromisoformat(log.check_out_time if log.check_out_time else log.check_in_time)
        time_diff = (now - last_action_time).total_seconds()
        
        if time_diff < 10: # 10 seconds cooldown
             print(f"Updates blocked: {time_diff}s since last scan")
             return {
                "action": "DUPLICATE_BLOCKED",
                "roll_no": clean_roll_no,
                "student_name": student.name,
                "status": log.status,
                "message": f"⏳ Please wait 10s before rescanning."
             }

        # === SECOND SCAN -> CHECK-OUT ===
        if log.check_out_time:
             # Already checked out - Update logic? 
             # For now, we allow updating checkout time ONLY if status is PENDING or ABSENT (re-evaluating)
             # But if PRESENT, we block.
             if log.status == "PRESENT":
                 return {
                    "action": "DUPLICATE_BLOCKED",
                    "roll_no": clean_roll_no,
                    "student_name": student.name,
                    "check_in_time": log.check_in_time,
                    "check_out_time": log.check_out_time,
                    "status": log.status,
                    "message": f"✅ {student.name} is already marked PRESENT."
                 }
        
        # Perform Check-out Calculation
        check_in_dt = datetime.fromisoformat(log.check_in_time)
        duration_minutes = (now - check_in_dt).total_seconds() / 60.0
        
        required_minutes = event.duration_minutes * (event.min_attendance_percent / 100.0)
        attendance_status = "PRESENT" if duration_minutes >= required_minutes else "ABSENT"
        
        # Prevent accidental rapid checkout (e.g. < 1 min)
        if duration_minutes < 1.0:
             return {
                "action": "EARLY_CHECKOUT_WARNING",
                "roll_no": clean_roll_no,
                "student_name": student.name,
                "message": f"⚠️ Too early to checkout! ({duration_minutes:.1f} min)"
             }

        log.check_out_time = now_iso
        log.duration_minutes = duration_minutes
        log.status = attendance_status
        
        db.commit()
        db.refresh(log)
        
        msg = f"✅ {student.name} — PRESENT ({duration_minutes:.1f} min)" if attendance_status == "PRESENT" else \
              f"❌ {student.name} — ABSENT ({duration_minutes:.1f} min < {required_minutes:.1f} min required)"

        return {
            "action": "CHECK_OUT",
            "roll_no": clean_roll_no,
            "student_name": student.name,
            "check_in_time": log.check_in_time,
            "check_out_time": now_iso,
            "duration_minutes": duration_minutes,
            "status": attendance_status,
            "message": msg
        }

File: server/routes/scan.js
const express = require('express');
const { getOne, runSql } = require('../db');

const router = express.Router();

// POST /api/scan — Core Attendance Engine
router.post('/', (req, res) => {
  try {
    const { roll_no, event_id } = req.body;

    if (!roll_no || !event_id) {
      return res.status(400).json({ error: 'roll_no and event_id are required' });
    }

    const cleanRollNo = String(roll_no).trim().toUpperCase();

    // Validate event
    const event = getOne('SELECT * FROM events WHERE id = ?', [Number(event_id)]);
    if (!event) {
      return res.status(404).json({ error: 'Event not found' });
    }
    if (!event.is_active) {
      return res.status(400).json({ error: 'Event is no longer active' });
    }

    // Validate student
    const student = getOne('SELECT * FROM students WHERE roll_no = ?', [cleanRollNo]);
    if (!student) {
      return res.status(404).json({
        error: 'Student not found',
        roll_no: cleanRollNo,
        message: `No student registered with roll number: ${cleanRollNo}`
      });
    }

    // Check existing attendance log
    const log = getOne('SELECT * FROM attendance_logs WHERE roll_no = ? AND event_id = ?', [cleanRollNo, Number(event_id)]);

    if (!log) {
      // ═══ FIRST SCAN → CHECK-IN ═══
      const now = new Date().toISOString();
      runSql(
        "INSERT INTO attendance_logs (roll_no, event_id, check_in_time, status) VALUES (?, ?, ?, 'PENDING')",
        [cleanRollNo, Number(event_id), now]
      );

      return res.status(201).json({
        action: 'CHECK_IN',
        roll_no: cleanRollNo,
        student_name: student.name,
        department: student.department,
        check_in_time: now,
        message: `✅ ${student.name} checked in successfully`
      });
    }

    if (!log.check_out_time) {
      // ═══ SECOND SCAN → CHECK-OUT + VALIDATE ═══
      const checkOut = new Date();
      const checkIn = new Date(log.check_in_time);
      const durationMs = checkOut.getTime() - checkIn.getTime();
      const durationMinutes = Math.round((durationMs / 60000) * 100) / 100;

      const requiredMinutes = event.duration_minutes * (event.min_attendance_percent / 100);
      const status = durationMinutes >= requiredMinutes ? 'PRESENT' : 'ABSENT';

      const checkOutISO = checkOut.toISOString();
      runSql(
        'UPDATE attendance_logs SET check_out_time = ?, duration_minutes = ?, status = ? WHERE roll_no = ? AND event_id = ?',
        [checkOutISO, durationMinutes, status, cleanRollNo, Number(event_id)]
      );

      return res.json({
        action: 'CHECK_OUT',
        roll_no: cleanRollNo,
        student_name: student.name,
        department: student.department,
        check_in_time: log.check_in_time,
        check_out_time: checkOutISO,
        duration_minutes: durationMinutes,
        required_minutes: requiredMinutes,
        status: status,
        message: status === 'PRESENT'
          ? `✅ ${student.name} — PRESENT (${durationMinutes.toFixed(1)} min)`
          : `❌ ${student.name} — ABSENT (${durationMinutes.toFixed(1)} min < ${requiredMinutes.toFixed(1)} min required)`
      });
    }

    // ═══ ALREADY COMPLETED → DUPLICATE BLOCKED ═══
    return res.status(409).json({
      action: 'DUPLICATE_BLOCKED',
      roll_no: cleanRollNo,
      student_name: student.name,
      status: log.status,
      message: `⚠️ ${student.name} has already checked in and out for this event`,
      check_in_time: log.check_in_time,
      check_out_time: log.check_out_time,
      duration_minutes: log.duration_minutes
    });

  } catch (err) {
    if (err.message && err.message.includes('UNIQUE constraint')) {
      return res.status(409).json({ error: 'Duplicate scan detected' });
    }
    console.error('Scan error:', err);
    res.status(500).json({ error: 'Failed to process scan' });
  }
});

module.exports = router;
